lib/view_component/collection.rb

```diff
@@ -0,0 +1,54 @@
+
+# frozen_string_literal: true
+
+module ViewComponent
+  class Collection
+    def initialize(component, object = nil, options)
+      @component = component
+      @object = object
+      @options = options
+    end
+
+    def render_in(view_context, &block)
+      as = as_variable(@component.virtual_path, @options)
+      collection = collection_variable(@object, @options)
+      args = @options.except(:collection, :as)
+
+      collection.map do |item|
+        args[as] = item
+        @component.new(args).render_in(view_context, &block)
+      end.join
+    end
+
+    private
+
+    def collection_variable(object, options)
+      if object.respond_to?(:to_ary)
+        object.to_ary
+      elsif options.key?(:collection)
+        collection = options[:collection]
+        collection ? collection.to_a : []
+      else
+        raise ArgumentError.new("Must specify the option `collection` or pass a valid collection object.")
+      end
+    end
```

Comment:
Is there a strong reason to have two ways to call the API? It's more code for us to maintain and a confusing API  (what would/should happen if I supply both options?) Given the user is calling a specific class method on the Component can we just force the first argument to be the collection?

Note that the Rails partial case is very different

```
<h1>Products</h1>
<%= render @products %>
```

This is a little bit of magic from Rails that says if you render an array it will treat it like a collection and infer the partial from the type of object in the array. We're not going to have that situation. The caller is going to pick the ViewComponent to call  `.all` on. 

The equivalent in ViewComponent world would be to support this:

```haml
render ViewComponent.all(@products) 
```

Under the covers ViewComponent would figure out the right ViewComponent class by type. In this case: `ViewComponent.all(@products) => ProductComponent.all(@products)`

However, that's a) maybe not necessary sugar for us b) a separate PR :)

Reply:
100% onboard with this.